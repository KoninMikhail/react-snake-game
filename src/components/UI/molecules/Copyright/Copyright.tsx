import { Text } from 'components/UI/atoms/body/Text/Text';
import { TextAlign } from 'components/UI/atoms/body/TextAlign/TextAlign';
import { StyledCopyright } from './StyledCopyright';

/**
 * @typedef ICopyrightProps
 *
 * @property {('left' | 'right' | 'center')} align - allowed to align values
 * @property {String} author - Project author name
 * @property {Number} creationYear - Project creation year
 * @property {String=} licence - Project Licence
 */

interface ICopyrightProps {
  align: 'left' | 'right' | 'center';
  author: string;
  creationYear: number;
  license?: string;
}

/**
 * Copyright
 *
 * @description Use for creation copyright label.
 *
 * @param {ICopyrightProps} props - Properties of copyright.
 * @returns {JSX.Element} Copyright element.
 */

export const Copyright = (props: ICopyrightProps): JSX.Element => {
  return (
    <StyledCopyright>
      <TextAlign align={props.align}>
        <Text as="div">{`Â© ${renderCopyrightDates(props.creationYear)}  |  ${
          props.author
        }. ${renderCopyrightLicenseText(props.license)}`}</Text>
      </TextAlign>
    </StyledCopyright>
  );
};

/**
 * renderCopyrightDates
 *
 * @description Use this function to generate dates in footer copyright template.
 *
 * @memberOf Copyright
 * @param {Number} startYear - project start year
 * @returns {String} - project dates string
 */

function renderCopyrightDates(startYear: number): string {
  const currentYear = new Date().getFullYear();

  if (startYear < currentYear) {
    return `${startYear}-${currentYear}`;
  }

  return currentYear.toString();
}

/**
 * renderCopyrightLicenseText
 *
 * @description Use for Render copyright license label.
 *
 * @memberOf Copyright
 * @param {string} licenseName - e.g. MIT, GPLv2 and etc.
 * @returns {string}
 */

function renderCopyrightLicenseText(licenseName?: string): string {
  return licenseName && licenseName.length > 0 ? licenseName + ' license.' : '';
}
