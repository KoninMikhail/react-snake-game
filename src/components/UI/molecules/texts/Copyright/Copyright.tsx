import { Text } from 'components/UI/atoms/body/Text/Text';
import { TextAlign } from 'components/UI/atoms/body/TextAlign/TextAlign';
import { StyledCopyright } from './StyledCopyright';
import calculateYearsDifference from '../../../../../utils/calculateYearsDifference';
import { useMemo } from 'react';

/**
 * @typedef ICopyrightProps
 *
 * @property {('left' | 'right' | 'center')} align - allowed to align values
 * @property {String} author - Project author name
 * @property {Number} creationYear - Project creation year
 * @property {String=} licence - Project Licence
 */
export interface ICopyrightProps {
  align?: 'left' | 'right' | 'center';
  author: string;
  creationYear: number;
  license: string;
}

/**
 * Copyright
 *
 * Use for creation copyright label.
 *
 * @param {ICopyrightProps} props - Properties of copyright.
 * @returns {JSX.Element} Copyright element.
 */
export const Copyright = ({ align = 'left', author, license, creationYear }: ICopyrightProps) => {
  const projectYears = useMemo(() => {
    return calculateYearsDifference(creationYear);
  }, [creationYear]);

  return (
    <StyledCopyright>
      <TextAlign align={align}>
        <Text as="div">{`Â© ${projectYears}  |  ${author}. ${license + ' license.'}`}</Text>
      </TextAlign>
    </StyledCopyright>
  );
};
