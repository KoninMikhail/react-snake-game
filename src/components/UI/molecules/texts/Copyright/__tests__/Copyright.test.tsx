import { Copyright } from '../Copyright';
import { render, screen } from '@testing-library/react';

describe('Copyright', () => {
  const currentYear = new Date().getFullYear();
  const prevYear = currentYear - 5;
  const license = 'GNUv2';
  const author = 'Mickey Mouse';

  it('renders copyright with current year', () => {
    const outputValue = `© ${currentYear} | ${author}. ${license} license.`;

    render(<Copyright author={author} creationYear={currentYear} license={license} />);

    expect(screen.getByText(/license/i)).toHaveTextContent(outputValue);
  });

  it('renders copyright with prev year', () => {
    const outputValue = `© ${prevYear}-${currentYear} | ${author}. ${license} license.`;

    render(<Copyright author={author} creationYear={prevYear} license={license} />);

    expect(screen.getByText(/license/i)).toHaveTextContent(outputValue);
  });

  it('render correctly', () => {
    const { container } = render(
      <Copyright align={'left'} author={author} creationYear={prevYear} license={license} />
    );
    expect(container).toMatchSnapshot();
  });
});
