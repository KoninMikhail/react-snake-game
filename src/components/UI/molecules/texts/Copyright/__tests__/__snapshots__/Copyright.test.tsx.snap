// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copyright render correctly 1`] = `
.c2 {
  font-size: 14px;
  font-family: 'pixeloid_bold','Segoe UI','Roboto','Oxygen','Ubuntu','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-weight: normal;
  color: white;
}

.c1 {
  text-align: left;
}

.c0 {
  display: block;
}

@media (min-width:1200px) {
  .c2 {
    font-size: 1vw;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Â© 2017-2022  |  Mickey Mouse. GNUv2 license.
      </div>
    </div>
  </div>
</div>
`;
